client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,696 @@
+};
+
+typedef struct {
+    uint32_t seq_num;
+    uint16_t type;// + reserved
+    uint16_t uid_length;
+    uint32_t size;
+    uint32_t topic_id;
+    void *uid_data;
+    void *notification_body;
+} kaa_notification_wrapper_t;
+
+typedef enum {
+    MANDATORY = 0x00,
+    OPTIONAL  = 0x01
+} notification_type_t;
+
+typedef enum {
+    ADD,
+    REMOVE
+} kaa_subscription_command_type_t;
+
+typedef struct {
+    uint32_t id;
+    kaa_subscription_command_type_t command;
+} kaa_subscription_command_t;
+
+typedef struct {
+    uint32_t topic_id;
+    kaa_list_t *listeners;
+} kaa_optional_notification_listeners_wrapper_t ;
+
+typedef struct {
+    uint32_t id;
+    kaa_notification_listener_t* listener;
+} kaa_notification_listener_wrapper_t;
+
+typedef struct {
+    uint32_t id;
+    kaa_topic_listener_t* listener;
+} kaa_topic_listener_wrapper_t;
+
+static kaa_service_t notification_sync_services[1] = {KAA_SERVICE_NOTIFICATION};
+
+static bool kaa_find_notification_listener_by_id(void *listener, uint32_t *context)
+{
+    return ((kaa_notification_listener_wrapper_t *) listener)->id == *context;
+}
+static bool kaa_find_topic_listener_by_id(void *listener, uint32_t *context)
+{
+    return ((kaa_topic_listener_wrapper_t *) listener)->id == *context;
+}
+
+static bool kaa_find_optional_notification_listener_by_id (void *optional_listener_list, uint32_t *topic_id)
+{
+    kaa_optional_notification_listeners_wrapper_t *optional_listener_node = (kaa_optional_notification_listeners_wrapper_t *) optional_listener_list;
+    return optional_listener_node->topic_id == *topic_id;
+}
+
+static bool kaa_find_topic_by_id(void *topic, uint32_t *id)
+{
+    kaa_topic_t* notification_topic = (kaa_topic_t *) topic;
+    return notification_topic->id == *id;
+
+}
+static void kaa_notification_wrapper_dealocator(void *data)
+{
+    kaa_optional_notification_listeners_wrapper_t * wrapper = (kaa_optional_notification_listeners_wrapper_t *) data;
+    KAA_FREE(wrapper);
+}
+static void kaa_notification_listener_deallocator(void *data)
+{
+    kaa_notification_listener_wrapper_t * wrapper = (kaa_notification_listener_wrapper_t *) data;
+    KAA_FREE(wrapper->listener);
+    KAA_FREE(wrapper);
+}
+
+static void kaa_topic_listener_deallocator(void *data)
+{
+    kaa_topic_listener_wrapper_t * listener = (kaa_topic_listener_wrapper_t  *) data;
+    KAA_FREE(listener->listener);
+    KAA_FREE(listener);
+}
+
+static void kaa_topic_deallocator(void* data)
+{
+    kaa_topic_t * topic = (kaa_topic_t *) data;
+    KAA_FREE(topic);
+}
+
+static void kaa_subscriptions_deallocator(void* data)
+{
+    kaa_subscription_command_t *cmd = (kaa_subscription_command_t *) data;
+    KAA_FREE(cmd);
+}
+static kaa_error_t kaa_find_topic(kaa_notification_manager_t *self, kaa_topic_t **topic, uint32_t *topic_id)
+{
+    KAA_RETURN_IF_NIL2(topic_id, topic, KAA_ERR_BADPARAM);
+    *topic = kaa_list_find_next(self->topics, &kaa_find_topic_by_id, topic_id);
```

Comment:
The kaa_list_find_next function returns the pointer to the found node in the list, but not the data in it.